### ISSUE TITLE
URGENT: Check if first requests are sent without authentication token

### ISSUE THREAD

**Issue by iamareebjamal** [fossasia/open-event-android#1006]
I was reading the implementation of Authenticator of OkHTTP3 and it seems it only calls the callback to set authentication headers IF the server fails and give 401. Please check if this is true since this can cause unauthenticated information to be shown to the user for endpoints which support both authenticated and unauthenticated users

@liveHarshit @GOVINDDIXIT Can you please check this?

**Comment by GOVINDDIXIT** [comment]
Yes, They have mentioned this in reactive authentication [here](https://square.github.io/okhttp/3.x/okhttp/okhttp3/Authenticator.html)
Here is our RequestAuthenticator

class RequestAuthenticator(private val authHolder: AuthHolder) : Authenticator {

    override fun authenticate(route: Route?, response: Response): Request? {
        val authorization: String? = authHolder.getAuthorization()
        return if (response.request().header("Authorization") == null && authorization != null) {
            response.request().newBuilder()
                    .header("Authorization", authorization)
                    .build()
        } else null // Give up, we already tried authenticating
    }
}

**Comment by iamareebjamal** [comment]
Then please change the implementation to use interceptor instead, as used in Orga App

**Comment by GOVINDDIXIT** [comment]
@liveHarshit you can work on this issue. I have looked into orga app. Since I am new to interceptor and need to research more about it before fixing.  Its an urgent issue and need to be fixed soon.

**Comment by liveHarshit** [comment]
> @liveHarshit you can work on this issue. I have looked into orga app. Since I am new to interceptor and need to research more about it before fixing. Its an urgent issue and need to be fixed soon.

Sure, but I will take some time. :+1:

**Comment by liveHarshit** [comment]
> Then please change the implementation to use interceptor instead, as used in Orga App

How to change module definition, because network module is giving error with Interceptor?

**Comment by iamareebjamal** [comment]
What error?

**Comment by liveHarshit** [comment]
> What error?

```
E/KOIN: [ERROR] - Error while resolving instance for class 'okhttp3.Authenticator' - error: org.koin.error.NoBeanDefFoundException: No compatible definition found. Check your module definition 
E/KOIN: [ERROR] - Error while resolving instance for class 'okhttp3.OkHttpClient' - error: org.koin.error.BeanInstanceCreationException: Can't create definition for 'Single [name='OkHttpClient',class='okhttp3.OkHttpClient']' due to error :
    		No compatible definition found. Check your module definition
    		org.koin.core.bean.BeanRegistry.checkedResult(BeanRegistry.kt:120)
    		org.koin.core.bean.BeanRegistry.retrieveDefinition(BeanRegistry.kt:87)
```

**Comment by iamareebjamal** [comment]
That's because you have not declared it in the module. Read docs of Koin

**Comment by liveHarshit** [comment]
> That's because you have not declared it in the module. Read docs of Koin

Yes, But I cannot find Implicit definitions of the module in the project. I have changed it as Interceptor in the network module. But Modules.kt is linking definitions with modules. Please help.

**Comment by iamareebjamal** [comment]
I couldn't understand? What do you mean by "But Modules.kt is linking definitions with modules"

**Comment by liveHarshit** [comment]
> I couldn't understand? What do you mean by "But Modules.kt is linking definitions with modules"

According to the documentation first, we declare Implicit definitions and then link it with modules, which is modules.kt.

**Comment by liveHarshit** [comment]
> you have not declared it in the module

But I cannot find module other then Modules.kt

**Comment by iamareebjamal** [comment]
There's no such need, just declare it in Modules.kt

**Comment by liveHarshit** [comment]
> There's no such need, just declare it in Modules.kt

Yes I have done it in network module - `single { RequestAuthenticator(get()) as Interceptor }` but after running the app - 


> E/KOIN: [ERROR] - Error while resolving instance for class 'okhttp3.Authenticator' - error: org.koin.error.NoBeanDefFoundException: No compatible definition found. Check your module definition E/KOIN: [ERROR] - Error while resolving instance for class 'okhttp3.OkHttpClient' - error: org.koin.error.BeanInstanceCreationException: Can't create definition for 'Single [name='OkHttpClient',class='okhttp3.OkHttpClient']' due to error : No compatible definition found. Check your module definition org.koin.core.bean.BeanRegistry.checkedResult(BeanRegistry.kt:120) org.koin.core.bean.BeanRegistry.retrieveDefinition(BeanRegistry.kt:87)

**Comment by iamareebjamal** [comment]
Because you haven't removed the authenticator argument from the OkHTTPClient Builder

**Comment by iamareebjamal** [comment]
https://github.com/fossasia/open-event-android/blob/f7b82dcf721d16806f5127e5d85f6985a1ce71e8/app/src/main/java/org/fossasia/openevent/general/di/Modules.kt#L158-L170

**Comment by iamareebjamal** [comment]
You have to remove the authenticator line
